In [2]:
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
%matplotlib inline
import scipy.stats as st
import seaborn as sns

import statsmodels.api as sm
from sklearn.metrics import  auc,  accuracy_score

In [3]:
data = pd.read_csv('sadiev.aa.csv', header = None)
data.head(5)

0         1         2         3         4         5         6         7   \
0  0.0 -0.300830  0.441077 -0.134191 -2.264717 -0.329407  1.715426 -0.436577   
1  1.0 -1.198002 -0.259217 -1.233736  0.806884 -0.857894 -0.122242 -0.807638   
2  1.0 -0.821442  0.512911 -0.444017 -1.364441  0.485729 -0.781654  0.878372   
3  0.0 -0.632765 -0.756107 -1.121652  0.117884 -0.274633  0.679934  0.099155   
4  0.0 -0.849300  0.505607  0.966130 -0.535571 -0.519688  2.083742  1.036304   

         8         9         10  
0  1.256920 -0.598929 -0.196800  
1 -0.920620 -0.876655  0.368266  
2 -0.391879 -0.078187 -0.084408  
3  0.460586  1.243536 -0.774501  
4 -2.525288 -1.240893  0.512203

In [4]:
X = data.values.T[1:]
y = data.values.T[0]
a = np.ones((X.shape[1],1))
X = np.concatenate((a.T , X), axis = 0)
X.shape, y.shape

((11, 1000), (1000,))

### Вопрос 1
Постройте модель логистической регрессии Y от X и свободного коэффициента. Каково отношение шансов Y=1 к событию f(X) = 1, где f(X) - отклик?

In [5]:
model0 = sm.Logit(y, X.T)
model0 = model0.fit()
model0.summary()

Optimization terminated successfully.
         Current function value: 0.686782
         Iterations 4


<class 'statsmodels.iolib.summary.Summary'>
"""
                           Logit Regression Results                           
==============================================================================
Dep. Variable:                      y   No. Observations:                 1000
Model:                          Logit   Df Residuals:                      989
Method:                           MLE   Df Model:                           10
Date:                Tue, 21 Apr 2020   Pseudo R-squ.:                0.008539
Time:                        15:24:12   Log-Likelihood:                -686.78
converged:                       True   LL-Null:                       -692.70
Covariance Type:            nonrobust   LLR p-value:                    0.2966
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0648      0.064     -1.011      0.312      -0.190       0.061
x1            -0.0419      0.063     -0.669      0.503      -0.165       0.081
x2             0.0293      0.064      0.456      0.648      -0.097       0.155
x3             0.0556      0.067      0.831      0.406      -0.076       0.187
x4             0.1035      0.064      1.627      0.104      -0.021       0.228
x5             0.0435      0.067      0.654      0.513      -0.087       0.174
x6             0.0589      0.062      0.952      0.341      -0.062       0.180
x7            -0.1153      0.064     -1.803      0.071      -0.241       0.010
x8             0.0645      0.066      0.979      0.328      -0.065       0.194
x9            -0.0046      0.062     -0.074      0.941      -0.126       0.117
x10            0.0883      0.062      1.420      0.156      -0.034       0.210
==============================================================================
"""

In [6]:
model = sm.GLM(y, X.T, family=sm.families.Binomial())
model = model.fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      y   No. Observations:                 1000
Model:                            GLM   Df Residuals:                      989
Model Family:                Binomial   Df Model:                           10
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -686.78
Date:                Tue, 21 Apr 2020   Deviance:                       1373.6
Time:                        15:24:13   Pearson chi2:                 1.00e+03
No. Iterations:                     4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0648      0.064     -1.011      0.312      -0.190       0.061
x1            -0.0419      0.063     -0.669      0.503      -0.165       0.081
x2             0.0293      0.064      0.456      0.648      -0.097       0.155
x3             0.0556      0.067      0.831      0.406      -0.076       0.187
x4             0.1035      0.064      1.627      0.104      -0.021       0.228
x5             0.0435      0.067      0.654      0.513      -0.087       0.174
x6             0.0589      0.062      0.952      0.341      -0.062       0.180
x7            -0.1153      0.064     -1.803      0.071      -0.241       0.010
x8             0.0645      0.066      0.979      0.328      -0.065       0.194
x9            -0.0046      0.062     -0.074      0.941      -0.126       0.117
x10            0.0883      0.062      1.420      0.156      -0.034       0.210
==============================================================================
"""

In [7]:
y_predict = model.predict(X.T)
OR = y.mean()/y_predict.mean()
print('Отношение шансов: {} '.format(round(OR, 2)))

Отношение шансов: 1.0 


### Вопрос 2

Какова точность (Accuracy) классификации модели при пороге вероятности для классификации p0 = 0.5 ?

In [8]:
y_predict = (model.predict(X.T) >= 0.5) * 1
print('Точность классификации: {}'.format(round(accuracy_score(y_predict , y), 2)))

Точность классификации: 0.56


### Вопрос 3

Постройте модель логистической регрессии Y от X и свободного коэффициента для первых 100 элементов выборки. Имеет ли смысл данная модель согласно критерию Вальда? (p-value=0.05)

In [9]:
y_new = y[:100]
X_new = X.T[:100]
model_new = sm.GLM(y_new, X_new, family=sm.families.Binomial())
model_new = model_new.fit()
model_new.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                 Generalized Linear Model Regression Results                  
==============================================================================
Dep. Variable:                      y   No. Observations:                  100
Model:                            GLM   Df Residuals:                       89
Model Family:                Binomial   Df Model:                           10
Link Function:                  logit   Scale:                          1.0000
Method:                          IRLS   Log-Likelihood:                -63.514
Date:                Tue, 21 Apr 2020   Deviance:                       127.03
Time:                        15:24:14   Pearson chi2:                     102.
No. Iterations:                     4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          z      P>|z|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0046      0.233     -0.020      0.984      -0.461       0.452
x1            -0.0663      0.195     -0.340      0.734      -0.449       0.316
x2             0.1289      0.235      0.548      0.584      -0.332       0.590
x3             0.0925      0.255      0.362      0.717      -0.408       0.593
x4             0.2941      0.216      1.360      0.174      -0.130       0.718
x5             0.1594      0.236      0.676      0.499      -0.303       0.621
x6            -0.1463      0.211     -0.693      0.488      -0.560       0.268
x7             0.0914      0.215      0.425      0.671      -0.330       0.513
x8             0.0279      0.204      0.136      0.891      -0.372       0.428
x9             0.3423      0.226      1.513      0.130      -0.101       0.786
x10            0.5378      0.248      2.171      0.030       0.052       1.023
==============================================================================
"""

In [10]:
model_new.wald_test('x1 = x2 = x3 = x4 = x5 = x6 = x7 = x8 = x9 = x10 = 0')

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[9.92070797]], p-value=0.4474772257870121, df_denom=10>

Мы получили, что $p_{\text{value}}=0.48 > 0.05$. Значит модель не имеет смысл согласно критерию Вальда.

### Вопрос 4

Можно ли обнулить первые четыре переменные согласно критерию Вальда? (p-value=0.05)

In [12]:
model_new.wald_test('x1 = x2 = x3 = x4 = 0')

<class 'statsmodels.stats.contrast.ContrastResults'>
<Wald test (chi2): statistic=[[2.44902383]], p-value=0.6537865294514112, df_denom=4>

Мы получили, что $p_{\text{value}}=0.65 > 0.05$. Значит можно обнулить первые четыре переменные согласно критерию Вальда.